### DISCUSSION TITLE
Is the "bottleneck" with client apps now?

### DISCUSSION THREAD

**Discussion by trosel** [markqvist/Reticulum#388] [General]
I was just thinking about this latest release 

> Reticulum can now establish links over even ultra low bandwidth mediums, all the way down to 5 bits per second.

The protocol continues to get better and better but adoption doesn't increase. Surely this is not because of shortcomings on the protocol side, right? Or is adoption kept low because the protocol is not ready for larger use yet?

When I think of how to use Reticulum from an average user perspective, I wouldn't even know where to start. The whole User Experience around "announce streams" confused me and I'm a fairly technical person. I know that Sideband is only just getting started on iOS, so my experience is probably skewed by that, but I would like to use a messaging app that uses Reticulum.

Should development shift gears more on the UX side of things for a bit to balance out the "flywheel" of adoption? More users could help to show the issues with the protocol more clearly which would lead to a better protocol, leading to more users, etc

**Comment by faragher** [comment]
There's a saying, "Build a better mousetrap, and the world will beat a path to your door." There is no better way to get successful businesspeople to laugh.

There is no "bottleneck" impeding an inevitable rise to success. For every successful project there are ten others that did the exact same thing but failed. Luck is a factor, but more than that, believing the "product sells itself" is a project killer.

There is a spectacular chain of better technology that lost to inferior products, and it almost always comes down to support from outside actors. The long, slow push towards acceptance is like pop-starting a car. You need a lot of initial energy to even have a chance of turning the engine over, but if you do it wrong, either it doesn't start or you cripple the vehicle. Five out of six Russian Roulette players will swear it's easy money.

I don't entirely agree with marketing agencies, their primary goal is to sell their services with a secondary concern of selling your product, but I do believe in the study of influence. Tom Cialdini is one of the best, or at least his descriptions are the most useful. In order to truly resonate with someone, you use reciprocity, consistency, social proof, authority, scarcity, and liking.

Reciprocity isn't really available here, but Reticulum is free and I like to think while it's not perfect, those of us supporting the project are cheerful, helpful, and generally more positive than much of the Internet. I'm not saying people should support Reticulum because we're putting in effort for the end user, but between that and Liking, I like to think we're on a good path to forming a solid community. The folk in Matrix chat are constantly sharing their experiences and helping us all move forward.

Commitment and Social Proof are big ones. If someone starts using the stack, they're invested in continuing to work with it, and if they see others using it, they'll be more likely to use it themselves. If the people they see using it remind them of themselves, that's even more important. And if those people appear to have authority, that's even more important.

So long as I have any say in it, Reticulum will never have a scarcity problem. Moving on.

So, what's my point? My point is the software applications aren't a problem, especially with the newest versions. The problem is the project applications. Many of us are working on that, but it's a delicate process. I spent two weeks on a single back of the envelope project assessment for a dam dredge, and while it was never actually performed, the local government seemed pleased both at my process and my conclusions. The amount of research and planning that went into a single study was massive, and this is what my undergrad had me doing for four years prior to graduate school. I knew the ins and outs of these proposals and how to properly care for and feed policymakers. 

A similar project for Reticulum can most certainly springboard us overnight, but it needs to be done right. Some local government or large organizations needs to see that this technology is at least worth a tech demo. For that to actually occur, and we'll only get a few shots at best, we need to show that we understand their problem, that we have a solution, and that it's worth changing their infrastructure for our solution. If you've ever worked with safety representatives, you know that "this is the right way" isn't worth a whole lot. You need to show the right project to the right person and frame it as "here's what I can do for you today, not tomorrow, not next week, today, and with this outlay right here, the next thing you'll get from us is a progress report and everyone will love you." And if you can't back that up in exactly the way they expect, they'll ignore you entirely.

Now, you may think this is ignoring the reason for Reticulum's creation. It is not. The recognition and media notice of such a project, indeed a string of projects in success, is exactly the bump needed to speed up widespread acceptance. Showing people the turn-key solution that everyone else has already worked the kinks out of is worth its weight in gold.

Simply showing this to people won't do us any good. Who wants secure communication? Not the average GMail or Yahoo mail user, that's for certain. There are technical people, idealists, pirates, criminals, and the paranoid. Sure, some folk dabble in secure communication, but the people willing to go out of their way are not the average user. And slow speeds really do reduce the appeal for pirates. The last time I saw a community pull for decentralized storage and secure communication it was FreeNet. FreeNet, by the end, was a repository of things that get other prisoners to shank you in prison. We must find a way to be useful to people that makes us a solid and upstanding option for communication, not just the newest Silk Road affiliate. 

Regarding shifting focus, Mark's newest Sideband is a thing of beauty. It is everything we need to make a dozen projects with great public interest possible. Search and Rescue, watching your kids at the zoo, surveying, mapping, remote weather stations, and so much more on a single app for a cheap phone hooked to an RNode. That's amazing. 

But the network stack itself is the important thing. Yes, it's feature complete as a networking stack, but every revision brings massive utility. This brings us to a proper allocation of resources.

There are only so many hours in a day, so many days in a life. There are plenty of developers who can make apps that link to a library. There is one person on this planet who knows the entire network stack front to back, along with its roadmap. Every hour spent on auxiliary projects is an hour not spent on the core stack. Since Sideband and NomadNet are open source, anyone can download, develop, and make a pull request. The entire point of Reticulum as it stands is that it allows people to make their own applications, and by centralizing them, we're doing it a disservice.

So, if people are interested in making an app that interfaces with Reticulum, they should. Adding Reticulum support to existing apps is also a great way to increase its usefulness, but you have to show people why they should switch over. "Oh, this is so much better" doesn't work. You need to find a person, see they need something, show them how Reticulum addresses that need, and then get them to support it. Then have people keep supporting it.

TCP didn't just spring up overnight, it was a long, hard road, and was thousands of dollars to acquire a license. Networking was a pain. Even with network cards, you needed to run a TSR for IPX or NetBIOS to actually do anything with your network. Microsoft basically rolled Winsock into 3.11WFW with limited success, and by shipping it with Windows 95, it suddenly became the easiest way to work with both Windows and BSD machines, from which the socket originated. Many, many protocols lie dead behind it, and even now the protocol is vastly different than its original feature set.

Regarding the difficulty in understanding Reticulum, it's no harder than TCP. You need to configure your IP address, making sure it doesn't conflict, set your subnet mask and your gateway, and if you don't want to type IPs all day, you'll need to find a DNS or run something like Windows Sharing. The routing between subnets gets pretty complicated, and there is NAT, PXE, TFTP, and a bunch of other issues that you need to know to run a fully functional network. It's boggling, and if you care about security, it's infuriating.

But you plug a device into a router, and DHCP does it all. Done. Log into a wireless router and you just need to know the passphrase (or nothing if you're on an unsecured network). Done!

These are very early days, and despite the rapid improvements and growth, you can't expect something like this to simply take over the world. It wouldn't surprise me if it was a decade before we knew if people will really accept Reticulum, so there's no need to be hasty. Find a need and fill it with an app. Make a beeper with an OLED for messages or weather reports. Set up a camera that sends home images when it sees a bird enter a birdhouse. Use them to track rental bikes. Find a niche and fill it.

One of my old bosses did the smart home thing and set it up to play the Imperial March whenever he entered a geofence around his home. Teslas do a light show and dance every Christmas. Do something small and neat, then go share it with Hackaday. Focus is where it needs to be, we just need more passionate individuals to make the ecosystem stronger. Mobile programmer? Make a messaging app more like what you think people would like. Technical writer? We all love documentation. RF Engineer? We could sure use you! Someone with no applicable skills? Love that you're interested. Tell a friend if you find it exciting.

Just be realistic in your expectations. This has been in the works for years and will be in the works for years more. It's not Athena, it won't burst forth fully grown. It's a city. Build a house, build a grocery store, build a road. Just keep building, don't worry about when we'll get a Trader Joe's. 

Not to go all Wolf of Wallstreet or Mike Rowe, but at some point we all need to remember, life is busy. It's noisy. We have limited time and bandwidth, That's why so many job interviews include "Sell me this pencil."

**Reply by trosel** [reply]
> My point is the software applications aren't a problem, especially with the newest versions. The problem is the project applications. 

I think there is a big market for people that would love to use a messaging app that runs on their own hardware, but without the pain of sysadmin work. This is basically:

1. buy 2 cheap RNode devices from @markqvist (give one to a friend?)
2. Install Sideband and put in the address of your RNode to talk over your network

If we just assume a simple messaging use-case, my question is how this can be improved today. And from my experience, the biggest issue seems to rely on the step 2 (the client app experience). Now this may not be a fault with the app, it could just be a fault with the instructions of how to use the app too. Not sure. But it's definitely not straight-forward.

I'm not saying @markqvist himself needs to divert resources to it, but maybe something could be done to make it easier for *other* client developers to help. For example, since Reticulum is written in Python, then the number frameworks for writing client apps is limited by that, right? If not, is there some special way that Reticulum could be used on some of the most popular frameworks like React Native, etc?

**Reply by faragher** [reply]
Running nodes in Sideband is in the pipe. Otherwise Sideband already takes care of identity generation, announces, etc. automatically. All you really need to do is point it to a propagation node which, as noted, is coming to Sideband in the near future. If the process is too difficult or unclear, document your particular process and issues, both to assist onboarding others and to find where there's a miscommunication in the documentation or startup tutorials.

There are already projects for porting Reticulum to other languages. Assisting any of those developers would expand the frameworks available. There's a C implementation that was making much more headway than my C# implementation, and if I recall there was talk of both Rust and Java. At some point we'll have to look at shared instances across languages, but I don't think that's a particular problem at this time.

There is, at this time, no way to use Reticulum without Python. You can use other languages to make calls to a custom Python middleman, but if you don't have Python on your target platform, you'll need to port it or assist in one of the porting projects.

**Reply by trosel** [reply]
> All you really need to do is point it to a propagation node which, as noted, is coming to Sideband in the near future

So as of right now, you can't use Sideband with your own RNode device? Meaning, everyone defaults to being on the same global network?

The only differentiating factor for me is that I could be messaging on a totally different sort of network. If I'm just messaging over the regular internet, I don't see a purpose in using Sideband. But if I can message over a network that I have a "stake" in (meaning, I am running it or partially running it with a node), then that is pretty cool.

**Reply by faragher** [reply]
> > All you really need to do is point it to a propagation node which, as noted, is coming to Sideband in the near future
> 
> So as of right now, you can't use Sideband with your own RNode device? Meaning, everyone defaults to being on the same global network?

As of right now you can't run without a computer running a propagation node. You can spin up a node right now on a computer or SBC and build your own network. If you want point to point, it may work if you just have RNodes on your phone with the same settings, but that's why you're having problems with announces and the like, because you want to avoid "the pain of sysadmin work" you won't get any network benefits because you haven't built any infrastructure. Once the nodes can run on an android device, you can act as your own network hub without any real effort on your part.

Is this not explained well enough in the documentation, or have you not read it?

**Reply by markqvist** [reply]
Just to make it clear, you can totally make 2 RNodes (or buy them from someone), install Sideband on you and your friends phone, and start messaging each other directly. That's one of the primary points. Having a system that works even with **zero** infrastructure. This works perfectly, and is basically as easy to set up as it can be.

No internet? No infrastructure? No nothing? Does it still work? Yes. Will it keep working? Yes.

If you then add one or more "Propagation Nodes", you suddenly have a fully fledged messaging system, that can store messages for disconnected/offline users, in a very resilient manner.

You can run your network completely isolated, or join it to a larger network. It's your choice. That's the beauty of the Reticulum protocol. You have a stake in the network and help out by setting up a propagation node in your area, or you can just use the ones that are being made available by other people.

**Reply by trosel** [reply]
> Just to make it clear, you can totally make 2 RNodes (or buy them from someone), install Sideband on you and your friends phone, and start messaging each other directly

that is what I was thinking too! Because this setting exists in the config. 

![IMG_9126](https://github.com/markqvist/Reticulum/assets/1299322/ece7ed99-eb21-4a77-8638-7129b143406e)

And theoretically, someone that is not super technical could do this fairly easily if there was a clear guide for it, right? 

One way that Sideband onboarding could improve I think is just helping users conceptually understand why they need to do things the way they do. What is an announce stream and why is it only necessary on Sideband and not on any other messaging app?

I know there is a way to make this sensible and clear in non-technical language. It just hasn’t been achieved yet with the current UX.

**Comment by cobraPA** [comment]
> 
> As of right now you can't run without a computer running a propagation node. You can spin up a node right now on a computer or SBC and build your own network. 

I haven't looked recently, but are you saying you can't enable propagation node on the Android client?  I thought this was a feature of Reticulum, not something that was run 'on a computer'?

(Now if it's not enabled for the Android client, then your statement is technically accurate.)

**Reply by markqvist** [reply]
Any LXMF client also contains the code to be a propagation node, Sideband just currently doesn't have a user interface switch for it. You can do it just fine on Android using `nomadnet` or just directly with `lxmd` running in Termux.
